// nodes.backup/moodleService old/moodleService.js
module.exports = function(RED) {
    function getData(config) {
        RED.nodes.createNode(this,config);
        var node = this;
        node.status({fill:"red",shape:"ring",text:"disconnected"});
        
        var fs = require('fs');
        
        var availableTabs;
        
        fs.readFile('node_modules/moodleService/availableTabs.json', 'utf8', function (err, data) {
			if (!err) {
				availableTabs = JSON.parse(data);
			}
			else {
				// manage errors
			}
		});
		
        node.on('input', function(msg) {
        	
        	msg2 = RED.util.cloneMessage(msg);
        	msg2.payload = false;
        	
        	if (msg.payload == true) {
        	
        		if (msg.topic == "/available") {
        			
        			msg.payload = availableTabs;
        			msg2.payload = true;
					msg2.topic = msg.topic;
					node.send([msg,msg2]);
        			
        		}
        		else {
        
					var mysql = require('mysql');
			
					var con = mysql.createConnection({
						host: "localhost",
						user: "fabiano",
						password: "m00dl3;",
						database: "moodle",
						port: 3307
					});
			
					con.on('error', function(err) {
						node.status({fill:"red",shape:"ring",text:"disconnected"});
					});

					con.connect(function(err) {
						if (!err) {
							node.status({fill:"green",shape:"dot",text:"connected"});
							var str = availableTabs.filter(x => x.url == msg.topic)[0].query;
							con.query(str, function (err, result) {
								if (!err) {
									msg.payload = result;
									msg2.payload = true;
									msg2.topic = msg.topic;
									node.send([msg,msg2]);
								}
								else {
									// manage errors
								}
							});
						}
						else {
							// manage errors
						}
					});
				
				}

			}
        });
    }
    RED.nodes.registerType("moodleService", getData);
}
